src/java.base/share/classes/java/io/SerializationMisdeclarationChecker.java

```diff
@@ -0,0 +1,259 @@
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.io;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Member;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.Arrays;
+
+import static jdk.internal.event.SerializationMisdeclarationEvent.*;
+import static java.lang.reflect.Modifier.*;
+
+final class SerializationMisdeclarationChecker {
+
+    private static final String SUID_NAME = "serialVersionUID";
+    private static final String SERIAL_PERSISTENT_FIELDS_NAME = "serialPersistentFields";
+    private static final String WRITE_OBJECT_NAME = "writeObject";
+    private static final String READ_OBJECT_NAME = "readObject";
+    private static final String READ_OBJECT_NO_DATA_NAME = "readObjectNoData";
+    private static final String WRITE_REPLACE_NAME = "writeReplace";
+    private static final String READ_RESOLVE_NAME = "readResolve";
+
+    private static final Class<?>[] WRITE_OBJECT_PARAM_TYPES = {ObjectOutputStream.class};
+    private static final Class<?>[] READ_OBJECT_PARAM_TYPES = {ObjectInputStream.class};
+    private static final Class<?>[] READ_OBJECT_NO_DATA_PARAM_TYPES = {};
+    private static final Class<?>[] WRITE_REPLACE_PARAM_TYPES = {};
+    private static final Class<?>[] READ_RESOLVE_PARAM_TYPES = {};
+
+    @SuppressWarnings("removal")
+    static void checkMisdeclarations(Class<?> cl) {
+        AccessController.doPrivileged((PrivilegedAction<Object>) () -> {
+            privilegedCheckSerialVersionUID(cl);
+            privilegedCheckSerialPersistentFields(cl);
+
+            privilegedCheckPrivateMethod(cl, WRITE_OBJECT_NAME,
+                    WRITE_OBJECT_PARAM_TYPES, Void.TYPE);
+            privilegedCheckPrivateMethod(cl, READ_OBJECT_NAME,
+                    READ_OBJECT_PARAM_TYPES, Void.TYPE);
+            privilegedCheckPrivateMethod(cl, READ_OBJECT_NO_DATA_NAME,
+                    READ_OBJECT_NO_DATA_PARAM_TYPES, Void.TYPE);
+
+            privilegedCheckAccessibleMethod(cl, WRITE_REPLACE_NAME,
+                    WRITE_REPLACE_PARAM_TYPES, Object.class);
+            privilegedCheckAccessibleMethod(cl, READ_RESOLVE_NAME,
```

Comment:
They are called `privilegedXXX` because they _are_ (already) privileged, not because they change the privileges.

But yes, in view of removing the security manager, the implementation could be more "modern". Will have a look.

Reply:
> https://github.com/openjdk/jdk/pull/17129/commits/ff1b7068bd902f3030267afbc468e3244c944604

Thanks - that looks much cleaner.